### PR TITLE
Add ADC module and incorporated into sensors module

### PR DESCRIPTION
**Pull request by stephen99scott**
# Summary of Changes
* Set up ADC1 for reading analog sensor data
* Confirmed functionality on the Nucleo
* Used a pot in place of the flex sensor for ease of changing the analog value
* Each ADC reading is two bytes and represents mV
* Below is in hexadecimal, and is a result of changing the resistance of the pot
Screenshot of data received
Nucleo connected to circuit with potentiometer

### CODE REVIEW

File: firmware/applications/SignBuddy/sensors.c [hunk 1]
```diff
@@ -23,11 +24,23 @@ static state_t state;
...
 }
 
+void tx_flex_data(flex_data_t *data)
+{
+  for (int i = 0; i < FLEX_SENSOR_CNT; i++) {
+    uint8_t tx_data[2] = { data->sensor_data[i] & 0xff, data->sensor_data[i] >> 8 };
+    for (int j = 0; j < 2; j++) {
+      ble_uart_tx(tx_data[j]);
+      uint32_t start_time = system_time_get();
+      while (system_time_cmp_ms(start_time, system_time_get()) < 100);
```
Comment (stephen99scott) [inline firmware/applications/SignBuddy/sensors.c:39]: Will modify `ble_uart_tx()` to be able to handle input other than uint8_t

Comment (adambujak) [reply firmware/applications/SignBuddy/sensors.c:39]: i think you should change it to `uint8_t *data, uint32_t length` if anything, and encoding should be handled in a higher level module. not for this pr or anything, but just in the future

**Comment by adambujak** [comment]
i looked it over quickly and it looks good. I would like to abstract the `ADC1` to the `board.h` file but I'll approve it so we get this merged

**Review by adambujak — APPROVED** [review]

File: firmware/applications/SignBuddy/adc.c [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  wait_loop_index = ((LL_ADC_DELAY_INTERNAL_REGUL_STAB_US * (SystemCoreClock / (100000 * 2))) / 10);
+  while (wait_loop_index != 0)
+    wait_loop_index--;
+
+  LL_ADC_StartCalibration(ADC1, LL_ADC_SINGLE_ENDED);
+
+  while (LL_ADC_IsCalibrationOnGoing(ADC1) != 0);
+
+  wait_loop_index = (ADC_DELAY_CALIB_ENABLE_CPU_CYCLES >> 1);
+  while (wait_loop_index != 0)
```
Comment (adambujak) [inline firmware/applications/SignBuddy/adc.c:68]: why is this needed?

File: firmware/applications/SignBuddy/adc.c [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  uint16_t ADC_Data_Raw   = VAR_CONVERTED_DATA_INIT_VALUE;
+  uint16_t ADC_Data_mVolt = 0;
+
+  LL_ADC_REG_StartConversion(ADC1);
+
+  while (LL_ADC_IsActiveFlag_EOC(ADC1) == 0);
+
+  LL_ADC_ClearFlag_EOS(ADC1);
+
+  ADC_Data_Raw = LL_ADC_REG_ReadConversionData12(ADC1);
```
Comment (adambujak) [inline firmware/applications/SignBuddy/adc.c:90]: nit: i don't like the case changes in the variable names i would much prefer `adc_data_raw`

File: firmware/applications/SignBuddy/adc.c [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+#define ADC_PREEMPT_PRIO                     0
+#define ADC_SUB_PRIO                         0
+
+#define VREF                                 ((uint32_t)5000)
+#define ADC_DELAY_CALIB_ENABLE_CPU_CYCLES    (LL_ADC_DELAY_CALIB_ENABLE_ADC_CYCLES * 32)
+#define VAR_CONVERTED_DATA_INIT_VALUE        (__LL_ADC_DIGITAL_SCALE(LL_ADC_RESOLUTION_12B) + 1)
+
+void adc_init(void)
+{
+  LL_ADC_InitTypeDef       ADC_InitStruct       = { 0 };
```
Comment (adambujak) [inline firmware/applications/SignBuddy/adc.c:14]: nit: yeah this naming sucks lol, I know this is the standard the sdk uses but let's not do it too haha

File: firmware/applications/SignBuddy/adc.c (hunk 3 above)
Comment (stephen99scott) [reply firmware/applications/SignBuddy/adc.c:90]: Yup will change

File: firmware/applications/SignBuddy/adc.c (hunk 4 above)
Comment (stephen99scott) [reply firmware/applications/SignBuddy/adc.c:14]: Changing

File: firmware/applications/SignBuddy/sensors.c (hunk 1 above)
Comment (stephen99scott) [reply firmware/applications/SignBuddy/sensors.c:39]: Yup sounds good, we can make that change when we integrate ble_uart with the logger

File: firmware/applications/SignBuddy/adc.c (hunk 2 above)
Comment (stephen99scott) [reply firmware/applications/SignBuddy/adc.c:68]: Calibration isn't necessary but it sounds like it can improve the accuracy of readings, see https://www.st.com/resource/en/application_note/cd00211314-how-to-get-the-best-adc-accuracy-in-stm32-microcontrollers-stmicroelectronics.pdf

File: firmware/applications/SignBuddy/adc.c [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  while (LL_ADC_IsActiveFlag_ADRDY(ADC1) == 0);
+}
+
+void ADC1_2_IRQHandler(void)
+{}
+
+uint16_t adc_read(void)
+{
+  uint16_t adc_data_raw = VAR_CONVERTED_DATA_INIT_VALUE;
+  uint16_t adc_data_mvolt = 0;
```
Comment (adambujak) [inline firmware/applications/SignBuddy/adc.c:82]: why declare these up here?

why declare adc_data_mvolt up here to 0? why not just declare it where it's set?

am i missing some functionality here

File: firmware/applications/SignBuddy/adc.c [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  adc_common_config.CommonClock = LL_ADC_CLOCK_SYNC_PCLK_DIV2;
+  LL_ADC_CommonInit(__LL_ADC_COMMON_INSTANCE(ADC1), &adc_common_config);
+
+  LL_ADC_REG_SetSequencerRanks(ADC1, LL_ADC_REG_RANK_1, LL_ADC_CHANNEL_9);
+  LL_ADC_SetChannelSamplingTime(ADC1, LL_ADC_CHANNEL_9, LL_ADC_SAMPLINGTIME_47CYCLES_5);
+  LL_ADC_SetChannelSingleDiff(ADC1, LL_ADC_CHANNEL_9, LL_ADC_SINGLE_ENDED);
+
+  LL_ADC_EnableIT_OVR(ADC1);
+
+  LL_ADC_DisableDeepPowerDown(ADC1);
```
Comment (adambujak) [inline firmware/applications/SignBuddy/adc.c:55]: should we add a define in the board file instead of using ADC1 in this file?

File: firmware/applications/SignBuddy/adc.c [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  GPIO_ADC_CLK_EN();
+
+  GPIO_InitStruct.Pin = ADC_PIN;
+  GPIO_InitStruct.Mode = LL_GPIO_MODE_ANALOG;
+  GPIO_InitStruct.Pull = LL_GPIO_PULL_NO;
+  LL_GPIO_Init(ADC_PORT, &GPIO_InitStruct);
+
+  LL_GPIO_EnablePinAnalogControl(ADC_PORT, ADC_PIN);
+
+  NVIC_SetPriority(ADC1_2_IRQn, NVIC_EncodePriority(NVIC_GetPriorityGrouping(), ADC_PREEMPT_PRIO, ADC_SUB_PRIO));
```
Comment (adambujak) [inline firmware/applications/SignBuddy/adc.c:29]: add a define for the irq too?

File: firmware/applications/SignBuddy/adc.c (hunk 2 above)
Comment (adambujak) [reply firmware/applications/SignBuddy/adc.c:68]: thanks

File: firmware/applications/SignBuddy/adc.c (hunk 6 above)
Comment (stephen99scott) [reply firmware/applications/SignBuddy/adc.c:55]: If you do that then you also need to abstract the IRQ function and I'm not sure how you would do that

File: firmware/applications/SignBuddy/adc.c (hunk 7 above)
Comment (stephen99scott) [reply firmware/applications/SignBuddy/adc.c:29]: See other reply, you would also need to abstract the IRQ callback, not sure how to do that

File: firmware/applications/SignBuddy/adc.c (hunk 5 above)
Comment (stephen99scott) [reply firmware/applications/SignBuddy/adc.c:82]: This doesn't seem like a big deal to me, but feel free to change it

Comment (stephen99scott) [reply firmware/applications/SignBuddy/adc.c:82]: The advantage is if the read fails then we will know because the result will be 0

File: firmware/applications/SignBuddy/adc.c (hunk 6 above)
Comment (adambujak) [reply firmware/applications/SignBuddy/adc.c:55]: unless I'm misunderstanding what you want, you can just add a #define MY_ADC_IRQ_Handler ADC1_IRQ_Handler

Comment (stephen99scott) [reply firmware/applications/SignBuddy/adc.c]: Thanks idk why I was under the impression you couldn't do that for functions, that's my bad. See f2bb66d
